Feature/3 rest api

kardan opened the pull request:
To keep the feedback cycle tight, testing was pushed to the next batch (since we don't yet have a testing strategy).

kardan committed:
Merge branch 'develop' into feature/3-rest-draft

kardan committed:
[#3] Initial REST API

- Added a proper processable workflow with a custom
  validator function and dispatching based on media type.
- Removed ring json in favour of later decoding that
  fits better with the liberator processable callbacks.
- Added generic not found strategy.
- Swaped media type supprot from clojure to edn.
- Added inital simple tests.
- Changed fixture data structure to match a
  typical db record.
- Added logic for not found in resources.
- Updated README to highlight that it's WIP.

iperdomo commented on src/akvo/notifications/data_store.clj:
```diff
@@ -16,27 +16,78 @@
+  {:id 2 :name \"Jane }}"
+  [tuple-vec]
+  (let [l (map #(assoc {} (keyword (str (:id %))) %) tuple-vec)]
+    (into {} l)))
```
I don't think you need to create a `let` and then use `into`, what about just using the `(map ...)` directly?

iperdomo replied:
A sample of this should something like:

``` clojure
 (reduce #(assoc %1 (keyword (str (:id %2))) %2) {} tuple-vec)
```

iperdomo commented on src/akvo/notifications/data_store.clj:
```diff
@@ -16,27 +16,78 @@
+(defn tuple-vec->id-tuple
+  "\"Vector of truples to tuples of id's\"; transforms a vector of
+  tuples into a map with the tuples id as keywords. Example: [{:id 1
+  :name \"Bob} {:id 2 :name \"Jane\"}] -> {:1 {:id :name \"example\"} :2
```
`{:1 {:id :name \"example\"}` should be `{:1 {:id 1 :name \"example\"}` ?

kardan replied:
Yes

iperdomo commented on src/akvo/notifications/rest_api.clj:
```diff
@@ -17,36 +17,155 @@
+  [ctx]
+  (slurp (get-in ctx [:request :body])))
+
+(defmulti #^{:private true} processable-body?
```
I think the whole logic of `processable-body?` is already  backed-in in liberator:
- https://github.com/clojure-liberator/liberator/blob/master/src/liberator/core.clj#L388

iperdomo replied:
You're right, on parsing the body of the request to get its content. `wrap-params` if for `application/x-www-form-urlencoded` body requests.

Perhaps is worth using this library instead of building it ourselves?
- https://github.com/ngrunwald/ring-middleware-format

kardan replied:
After discussions concluded that we need to handle processable logic ourselves in the Liberator layer. This has to do with the Liberator decision flow.

iperdomo commented on src/akvo/notifications/rest_api.clj:
```diff
@@ -17,36 +17,155 @@
+  {:pre [(:name data)
+         (not (contains? (->> @data/services-data
+                              (map #(assoc {} (:name %) (:id %)))
+                              (into {}))
```
Perhaps using `reduce` instead of `map` and then `(into {})` ?

kardan committed:
[#3 #6] Updated REST API after review

- Fixed code after review.
- Seperated rest-api into rest-api and rest-utils namespaces.
- Added simple tests using Midje.

kardan committed:
[#3 #6] Updates after code review

- Removed jackson as a dependency.
- Removed commas that have sneaked in.

iperdomo commented:
:+1: code reviewed
